Servers/UI/OJS.Servers.Ui/ClientApp/src/components/contests/contest-categories/ContestCategories.tsx

```diff
@@ -1,37 +1,82 @@
-import React, { useCallback } from 'react';
-import { isEmpty, isNil } from 'lodash';
+import React, { useCallback, useMemo } from 'react';
+import { isNil } from 'lodash';
 import { useContestCategories } from '../../../hooks/use-contest-categories';
 import Heading, { HeadingType } from '../../guidelines/headings/Heading';
 
 import styles from './ContestCategories.module.scss';
 import { useContests } from '../../../hooks/use-contests';
 import { IHaveOptionalClassName } from '../../common/Props';
 import Tree, { ITreeItemType } from '../../guidelines/trees/Tree';
+import { IFilter } from '../../../common/contest-types';
 
 interface IContestCategoriesProps extends IHaveOptionalClassName {
+    onCategoryClick: (filter: IFilter) => void;
+    defaultSelected?: string,
 }
 
-const ContestCategories = ({ className = '' }: IContestCategoriesProps) => {
+const ContestCategories = ({
+    className = '',
+    onCategoryClick,
+    defaultSelected,
+}: IContestCategoriesProps) => {
     const { state: { categories } } = useContestCategories();
+    const { state: { possibleFilters } } = useContests();
 
-    const {
-        state: { possibleFilters },
-        actions: { applyFilter },
-    } = useContests();
-
-    const handleTreeItemClick = useCallback((node: ITreeItemType) => {
-        if (!isEmpty(node.children)) {
-            return;
-        }
-
+    const handleTreeLabelClick = useCallback((node: ITreeItemType) => {
         const filter = possibleFilters.find(({ value }) => value.toString() === node.id.toString());
 
         if (isNil(filter)) {
             return;
         }
 
-        applyFilter(filter, true);
-    }, [ applyFilter, possibleFilters ]);
+        onCategoryClick(filter);
+    }, [ possibleFilters, onCategoryClick ]);
+
+    const flattenTree = useCallback(
+        (treeItems: ITreeItemType[], result: ITreeItemType[]) => {
+            treeItems.forEach(({ children, ...rest }) => {
+                result.push(rest);
+                if (!isNil(children)) {
+                    flattenTree(children, result);
+                }
+            });
+            return result;
+        },
+        [],
+    );
```

Comment:
Can't `result` be made an optional param? This way the first (external) call will not provide an empty list.

Reply:
If it optional I should create new array on every recursive call, by copying the provided one, in order to handle the missing array on the first iteration.